test-suite/src/main/scala/testsuite/core/IsInstanceOfTest.scala

```diff
@@ -50,6 +52,51 @@ object IsInstanceOfTest {
     !base2.isInstanceOf[Base]
   }
 
+  private def testArrays(): Unit = {
+    val child: Any = new Child
+    val booleanArray: Any = new Array[Boolean](1)
+    val intArray: Any = new Array[Int](1)
+    val stringArray: Any = new Array[String](1)
+    val intArrayArray: Any = new Array[Array[Int]](1)
+    val stringArrayArray: Any = new Array[Array[String]](1)
+    val objectArray: Any = new Array[Object](1)
+
+    // direct instances
+    ok(booleanArray.isInstanceOf[Array[Boolean]])
+    ok(intArray.isInstanceOf[Array[Int]])
+    ok(stringArray.isInstanceOf[Array[String]])
+    ok(intArrayArray.isInstanceOf[Array[Array[Int]]])
+    ok(stringArrayArray.isInstanceOf[Array[Array[String]]])
+    ok(objectArray.isInstanceOf[Array[Object]])
+
+    // primitive arrays are not instances of any other type
+    ok(!booleanArray.isInstanceOf[Array[Int]])
+    ok(!booleanArray.isInstanceOf[Child])
+    ok(!booleanArray.isInstanceOf[Array[Array[Boolean]]])
+    ok(!booleanArray.isInstanceOf[Array[Object]])
+
+    // different dimensions
+    ok(!intArray.isInstanceOf[Array[Array[Int]]])
+    ok(!intArrayArray.isInstanceOf[Array[Int]])
+    // TODO ok(!intArrayArray.isInstanceOf[Array[Array[Array[Int]]]])
+    // TODO ok(!stringArrayArray.isInstanceOf[Array[String]])
+    // TODO ok(!stringArrayArray.isInstanceOf[Array[Array[Array[String]]]])
+    // TODO ok(!objectArrayArray.isInstanceOf[Array[Array[Object]]])
+
+    // reference array types are covariant
```

Comment:
Sorry if I misunderstanding, but I'm a bit confused. 
AFAIK, both Scala's Array and WasmGC's array are invariant (if it's mutable per WasmGC).
but you're commenting that `reference array types` are covariant, and indeed it seems like so: when `A <: B`, `x: Array[A].isInstanceOf[Array[B]]` is true (on JVM), I thought it's going to be `false` 🤔 

and the compiler says something like

```scala
class Box[A](var content: A)
abstract class Animal {
  def name: String
}
case class Cat(name: String) extends Animal
case class Dog(name: String) extends Animal

val myCatBox: Box[Cat] = new Box[Cat](Cat("Felix"))

scala> myCatBox.isInstanceOf[Box[Animal]]
1 warning found
-- Unchecked Warning: ----------------------------------------------------------
1 |myCatBox.isInstanceOf[Box[Animal]]
  |^^^^^^^^
  |the type test for Box[Animal] cannot be checked at runtime because its type arguments can't be determined from Box[Cat]
val res25: Boolean = true
```

I thought it's going to be `false`, instead of `true`.
This is `true` on JVM because of type erasure (kinda same reason as WasmGC) or it's more like Scala's spec?

Reply:
It's not type erasure. It's the spec of JVM arrays at run-time. Scala follows that spec for run-time type tests, even though at compile-time its `Array` is indeed invariant.

In practice, lot of code relies on that property. In particular, there is a lot of code that relies on `Array[A] instanceof Array[Object]` for *every* reference type `A` (including arrays). So in the spec of the Scala.js IR, we also made its `ArrayType`s covariant for reference types.

When compiling to JS or Wasm, we then have to follow that spec.

Reply:
Ah, that makes sense, thank you for your expalanation!